### PR TITLE
Add workarounds for building against CVC4 on ArchLinux.

### LINKED ISSUE #4762: Build failures with CVC4, e.g. on ArchLinux
We have several independent reports about solidity failing to build against CVC4 on current Archlinux.

### PR DESCRIPTION
**Pull request by ekpyron**
Fixes #4762.

### CODE REVIEW

File: libsolidity/formal/CVC4Interface.h [hunk 1]
```diff
@@ -21,8 +21,16 @@
 
 #include <boost/noncopyable.hpp>
 
+#if defined(__GLIBC__)
+#define _GLIBCXX_PERMIT_BACKWARD_HASH
```
Comment (ekpyron) [inline libsolidity/formal/CVC4Interface.h]: This is ugly, but, unfortunately, I haven't found a way around it.

File: cmake/FindCVC4.cmake [hunk 2]
```diff
@@ -1,4 +1,12 @@
 find_path(CVC4_INCLUDE_DIR cvc4/cvc4.h)
-find_library(CVC4_LIBRARY NAMES cvc4 )
+find_library(CVC4_LIBRARY NAMES cvc4)
 include(FindPackageHandleStandardArgs)
 find_package_handle_standard_args(CVC4 DEFAULT_MSG CVC4_LIBRARY CVC4_INCLUDE_DIR)
+if(CVC4_FOUND)
+    find_library(CVN_LIBRARY NAMES cvn)
+    if(CVN_LIBRARY)
+        set(CVC4_LIBRARIES ${CVC4_LIBRARY} ${CVN_LIBRARY})
```
Comment (axic) [inline cmake/FindCVC4.cmake]: What is CVN?

**Comment by mattaereal** [comment]
The tests are failing due to download issues, not because of the commits. I'm currently testing this on my Arch Linux.

File: cmake/FindCVC4.cmake (hunk 2 above)
Comment (ekpyron) [reply cmake/FindCVC4.cmake]: It's CLN - I must have garbled the name in the latest rebase, not sure how that happened.

**Comment by ekpyron** [comment]
The new library is https://www.ginac.de/CLN/ - that seems to be an optional dependency of CVC4 and linking fails with undefined references to it, since CVC4 seems to require CLN in its default build in Archlinux.

**Comment by ekpyron** [comment]
@mattaereal I just changed this PR again, because I somehow garbled the name, but now it indeed seems to work for me on Archlinux.

File: libsolidity/formal/CVC4Interface.h (hunk 1 above)
Comment (chriseth) [reply libsolidity/formal/CVC4Interface.h]: What does it do? Can you add a comment?

**Comment by mattaereal** [comment]
No problem, building again, will let you know in a few minutes.

File: libsolidity/formal/CVC4Interface.h (hunk 1 above)
Comment (ekpyron) [reply libsolidity/formal/CVC4Interface.h]: I added a comment.

**Comment by mattaereal** [comment]
It just built the object for the solvers interfaces without an issue, and that's where it was halting, I don't think it will be failing from this point.

**Comment by mattaereal** [comment]
Confirmed. Built successful. From all Arch Linux users, thanks!

**Comment by axic** [comment]
@ekpyron does the test suite work properly after this change with cvc4 only on your machine?

**Comment by mattaereal** [comment]
Now this merge conflicts with #4665 that was about to get merged.

**Comment by ekpyron** [comment]
@axic I'll check (haven't checked CVC4 only so far, but it makes sense to do).

As some further explanation for this change:

CVC4 generates a header from https://github.com/CVC4/CVC4/blob/master/src/util/rational.h.in, in which it is hard-coded whether CLN was enabled at build time.
Then in https://github.com/CVC4/CVC4/blob/master/src/util/rational_cln_imp.h it wraps CLN directly in the header, so solidity ends up using CLN and has to link against it.

Is there actually any reason why solidity is linked against ``GMP`` other than CVC4? Because GMP is *not* needed, if CVC4 was built against CLN instead, so if CVC4 is the only reason, we might want to change the logic a bit more.

**Comment by ekpyron** [comment]
Yes, following the SMT/CVC4 logic in the CMake files, GMP is also only added, because CVC4 depends on it. I'll change this PR further to work well with both cases (CVC4+GMP and CVC4+CLN).

**Comment by mattaereal** [comment]
Just ran `scripts/tests.sh` and did not fail. But I can't remember if that tests for solvers too.

**Comment by ekpyron** [comment]
I just added another commit that removes requiring GMP as prerequisite of CVC4, since CVC4 built against CLN doesn't require GMP.

**Comment by ekpyron** [comment]
@axic I just ran the tests while disabling Z3 and only enabling CVC4 and they pass.

**Comment by ekpyron** [comment]
@mattaereal Thanks for confirming! Regarding the conflict with #4665: I will take care of it and rebase the second PR after the first is merged.

**Review by chriseth — DISMISSED** [review]

**Comment by mattaereal** [comment]
@ekpyron Awesome!

**Comment by axic** [comment]
I think this looks good, lets just merge the other one first.

**Comment by axic** [comment]
Please rebase after #4665.

**Comment by ekpyron** [comment]
@axic Rebased.

File: cmake/FindCVC4.cmake [hunk 3]
```diff
@@ -1,8 +1,26 @@
...
     find_path(CVC4_INCLUDE_DIR cvc4/cvc4.h)
-    find_library(CVC4_LIBRARY NAMES cvc4 )
+    find_library(CVC4_LIBRARY NAMES cvc4)
     include(FindPackageHandleStandardArgs)
     find_package_handle_standard_args(CVC4 DEFAULT_MSG CVC4_LIBRARY CVC4_INCLUDE_DIR)
+    if(CVC4_FOUND)
+        # CVC4 may depend on either CLN or GMP.
+        # We can assume that the one it requires is present on the system,
+        # so we quietly try to find both and link against them, if they are
+        # present.
```
Comment (axic) [inline cmake/FindCVC4.cmake:10]: I'm fine with this, but likely we may run into a compilation issue reported by someone where this isn't the case. Think about statically linked CVC4 against GMP?

Comment (axic) [reply cmake/FindCVC4.cmake:10]: Discussed this on gitter. Resolution: 
> Perhaps we should include that in the changelog: should you encounter issues with the experimental z3 / cvc4 support, please disable them via `USE_Z3=OFF USE_CVC4=OFF` when configuring cmake.

Comment (ekpyron) [reply cmake/FindCVC4.cmake:10]: Added some Changelog entry (I'm open to suggestions).

**Review by axic — APPROVED** [review]

**Pull request merged**